### PR TITLE
Rework Sheet 8: Multilayer Perceptron

### PR DESCRIPTION
**Pull request by ahoereth**
This is a complete rework of sheet 08 as proposed in #1. Let's get into reviewing it, the sheet is coming up shortly in the 2017 iteration of the class. Tracking some final questions in this PR.

### CODE REVIEW

File: sheet08/sheet08solutions.ipynb [hunk 1]
```diff
@@ -29,294 +29,231 @@
...
+    "                In a network of perceptrons this basically represents the \n",
+    "                number of neurons in the layer before this neuron's layer. \n",
+    "                Used for generating the perceptron's weights vector, which \n",
+    "                not only includes one weight per input but also an additional \n",
+    "                bias weight.\n",
+    "            act_fun: Function to apply on activation.\n",
+    "        \"\"\"\n",
+    "        self.act_func = act_func\n",
+    "        # self.weights = np.array([])  # TODO\n",
+    "        self.weights = np.random.normal(size=dim_in + 1)\n",
```
Comment (ahoereth) [inline sheet08/sheet08solutions.ipynb:109]: The lecture apparently does not explicitly talk about biases. We implement them using an additional weight, which is kind of intuitive but simplifies applying it to the inputs greatly. Might need further explanations, maybe in the practice session? @krumnack

File: sheet08/sheet08solutions.ipynb [hunk 2]
```diff
@@ -327,373 +264,223 @@
...
     "\n",
-    "    def __repr__(self):\n",
-    "        return '{}({}, {})'.format(self.__class__.__name__, len(self.w) - 1, self.epsilon)"
+    "        # Backpropagate error through the network computing\n",
+    "        # intermediate delta and adapting each layer.\n",
+    "        for k in range(len(self.layers) - 1, 0, -1):\n",
+    "            delta = (outputs[k + 1] * (1 - outputs[k + 1])) * e\n",
+    "            weights = self.layers[k].get_weights_matrix()\n",
+    "            e = (weights @ delta)[1:]\n",
+    "            self.layers[k].adapt(outputs[k], delta, rate)"
```
Comment (ahoereth) [inline sheet08/sheet08solutions.ipynb]: Not quite happy with this loop. The `k` and `k+1` usage is inconsistent with the formula given in the explanation above the code cell.

The previous solution was nicer, but uses a zip over three lists. Advantage here is that x and y explicitly hint at "this is the layers in- and output".

```python
layers = list(zip(self.layers, outputs[:-1], outputs[1:]))
for layer, x, y in reversed(layers):
    delta = (y * (1 - y)) * e
    weights = layer.get_weights_matrix()
    e = (weights @ delta)[1:]
    layer.adapt(x, delta, rate)
```

Comment (shoeffner) [reply sheet08/sheet08solutions.ipynb]: You could make it more explicit by assigning new variables:

```python
for k in range(len(self.layers) - 1, 0, -1):
    layer_output = outputs[k + 1]
    layer_input = outputs[k]
    delta = (layer_output * (1 - layer_output)) * e
    weights = self.layers[k].get_weights_matrix()
    e = (weights @ delta)[1:]
    self.layers[k].adapt(layer_input, delta, rate)
```

File: sheet08/sheet08solutions.ipynb [hunk 3]
```diff
@@ -327,373 +264,223 @@
...
+    "\n",
+    "In the cell below we visualize the data set."
    ]
   },
   {
    "cell_type": "code",
    "execution_count": null,
    "metadata": {
-    "collapsed": false
+    "collapsed": true
```
Comment (shoeffner) [inline sheet08/sheet08solutions.ipynb]: I just blank out all metadata where possible, but I think `collapsed` for a visualization cell is not always a good idea, even though it's just one click.

File: sheet08/sheet08solutions.ipynb [hunk 4]
```diff
@@ -29,294 +29,231 @@
...
+    "            rate: Learning rate.\n",
+    "        \"\"\"\n",
+    "        # self.weights += 0  # TODO\n",
+    "        self.weights += rate * delta * np.append(1, x)\n",
+    "\n",
+    "\n",
+    "_p = Perceptron(2)\n",
+    "assert _p.weights.size == 3, \"Should have a weight per input and a bias.\"\n",
+    "assert isinstance(_p.activate([2, 1]), float), \"Should activate as scalar.\"\n",
+    "assert -1 <= _p.activate([100, 100]) <= 1, \"Should activate using sigmoid.\""
```
Comment (shoeffner) [inline sheet08/sheet08solutions.ipynb]: To assert the adapt function, you could try something like:

```python
_p.weights = np.array([.5, .5, .5])
_p.adapt(np.array([2, 3]), np.array([.5, .5, .5]))
assert np.allclose(_p.weights, np.array([0.5075, 0.515, 0.5225])), "The weights are not updated properly."
```

File: sheet08/sheet08solutions.ipynb (hunk 3 above)
Comment (ahoereth) [reply sheet08/sheet08solutions.ipynb]: Notebooks are bad for version control. Fixed in 6e42864.

File: sheet08/sheet08solutions.ipynb (hunk 2 above)
Comment (ahoereth) [reply sheet08/sheet08solutions.ipynb]: Still does not tackle the problem with the weird `k` and `k+1` when comparing to the formula. Its basically the wrong way around. Maybe we could adjust the formula accordingly?

Comment (shoeffner) [reply sheet08/sheet08solutions.ipynb]: How about:

```python
for k, layer in reversed(list(enumerate(self.layers, 1))):
    layer_output = outputs[k]
    layer_input = outputs[k - 1]
    delta = (layer_output * (1 - layer_output)) * e
    weights = layer.get_weights_matrix()
    e = (weights @ delta)[1:]
    layer.adapt(layer_input, delta, rate)
```

File: sheet08/sheet08solutions.ipynb (hunk 4 above)
Comment (ahoereth) [reply sheet08/sheet08solutions.ipynb]: Does not work for me with those values...? Working for me:
```
_p = Perceptron(2)
_p.weights = np.array([.5, .5, .5])
_p.adapt(np.array([2, 3]), np.array([.5, .5, .5]))
assert np.allclose(_p.weights, [0.515, 0.53, 0.545]), "Should update weights."
```

Comment (ahoereth) [reply sheet08/sheet08solutions.ipynb]: Fixed in 364c9ad.

**Review by NinaWie — APPROVED** [review]

**Review by shoeffner — APPROVED** [review]

**Comment by ahoereth** [comment]
Zum mergen dieses branches in den SS17 master @krumnack:

```
git remote add public https://github.com/CogSciUOS/MachineLearningSheets.git
git fetch public -a
git checkout master
git merge public/SS17/sheet08
```

**Comment by krumnack** [comment]
Amazing improvement of sheet 08.

**Comment by ahoereth** [comment]
Would love to read your answer to the full-text question regarding the training performance :)

**Pull request merged**